pkg/logql/count_min_sketch.go

```diff
@@ -174,21 +177,23 @@ func NewHeapCountMinSketchVector(ts int64, metricsLength, maxLabels int) HeapCou
 			F:       f,
 			Metrics: make([]labels.Labels, 0, metricsLength+1),
 		},
-		observed:  make(map[string]struct{}),
+		observed:  make(map[uint64]struct{}),
 		maxLabels: maxLabels,
+		buffer:    make([]byte, 0, 1024),
 	}
 }
 
 func (v *HeapCountMinSketchVector) Add(metric labels.Labels, value float64) {
-	// TODO: we save a lot of allocations by reusing the buffer inside metric.String
-	metricString := metric.String()
-	v.F.Add(metricString, value)
+	v.buffer = metric.Bytes(v.buffer)
+
+	v.F.Add(v.buffer, value)
 
 	// Add our metric if we haven't seen it
-	if _, ok := v.observed[metricString]; !ok {
+	id := xxhash.Sum64(v.buffer)
```

Comment:
there's already a hash function for `labels.Labels` we could use? I suppose what you're doing is actually better, since if we called `metric.Bytes` and `metric.Hash` we're iterating over the whole slice of labels twice

the only difference is prometheus' labels hash function limits the total size of the bytes used for the hash https://github.com/prometheus/prometheus/blob/main/model/labels/labels.go#L76-L87

Reply:
>prometheus' labels hash function limits the total size of the bytes used for the hash

Not quite. It grows the bytes slice if it's not big enough. Unfortunately the function won't let one reuse the byte slice to avoid allocations. That's what I've found doing https://github.com/grafana/loki/commit/efac69032854e9f0e31aa2133caf9141e9c3c968